### ISSUE TITLE
Version cookie attribute added to the Set-Cookie header causes cookie parsing failures in a Reverse Proxy appliance.

### ISSUE THREAD

**Issue by thomasdarimont** [quarkusio/quarkus#44364]
### Description

In a customer project, we had issues with a quarkus application (Keycloak) where a commercial reverse proxy appliance would partially remove the cookies set via the set-cookie header (all attributes except cookie name and value attribute were removed). After investigating the problem, we learned that the root cause of the
failure was the `Version=1;` cookie attribute, which caused the reverse proxy applies to fail.

I know that in this case, it is probably the fault of the reverse proxy appliance for not parsing cookie values in a more robust way,
but since the product claims to follow the RFC 6265 standard, I wondered whether the current quarkus implementation anticipates a too-old RFC 2965, and obsolete, set-cookie handling.

The current implementation of [NewCookieHeaderDelegate](https://github.com/quarkusio/quarkus/blob/main/independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/headers/NewCookieHeaderDelegate.java#L59
) uses the `Set-Cookie2` defined by [RFC 2965 HTTP State Management Mechanism](https://datatracker.ietf.org/doc/html/rfc2965) from October 2000 
but for the `Set-Cookie` header.

`RFC 2965  HTTP State Management Mechanism` defines the `Set-Cookie2` syntax in 
```
 3.2.2  Set-Cookie2 Syntax
 
  Version=value
      REQUIRED. The value of the Version attribute, a decimal integer,
      identifies the version of the state management specification to
      which the cookie conforms. For this specification, Version=1
      applies.
```

This RFC is Obsoleted by [RFC 6265 HTTP State Management Mechanism](https://datatracker.ietf.org/doc/html/rfc6265) (April 2011, RFC - Proposed Standard)

In RFC the syntax for the Set-Cookie header is defined in  [4.1.1.  Syntax](https://www.rfc-editor.org/rfc/rfc6265#section-4.1.1)

The "Version" attribute is no longer listed there. 

I propose to make it possible to skip sending the `Version` attribute in the Set-Cookie header.

In the concrete customer project I had to work around this issue, by shadowing the `org.keycloak.quarkus.runtime.integration.resteasy.QuarkusHttpResponse`
and replace the code in `org.keycloak.quarkus.runtime.integration.resteasy.QuarkusHttpResponse#setCookieIfAbsent`
with the following:
```java
    @Override
    public void setCookieIfAbsent(NewCookie newCookie) {
        if (newCookie == null) {
            throw new IllegalArgumentException("Cookie is null");
        }

        if (newCookies == null) {
            newCookies = new HashSet<>();
        }

        if (newCookies.add(newCookie)) {
            String headerValue = NEW_COOKIE_HEADER_DELEGATE.toString(newCookie);

            // PATCH:BEGIN
            // Remove Version=1; fragment from set-cookie response header line.
            headerValue = headerValue.replace("Version=1;","");
            // PATCH:END

            addHeader(HttpHeaders.SET_COOKIE, headerValue);
        }
    }
```

### Implementation ideas

Perhaps it would be possible to check a quarkus configuration property in `org.jboss.resteasy.reactive.common.headers.NewCookieHeaderDelegate` to use old "RFC 2965" or new "RFC 6265" style cookies (the latter without the Version attribute).
Or have a specific property like `...cookies.add-version` with the default `true.`

**Comment by geoand** [comment]
@FroMage is the expert at reading and interpreting specs :)

**Comment by FroMage** [comment]
I already noticed that we were using outdated specs for cookies. Even the javadoc for `NewCookie` doesn't mention anything besides the outdated RFC and version 1.

What I'm amazed at is that nobody reported this until now.

Honestly, if it's as simple as dropping the version field, let's go for it. I assume the TCK will fail, but it's outdated anyway.

**Comment by thomasdarimont** [comment]
> What I'm amazed at is that nobody reported this until now.
If customers don't have to use "limited" reverse proxy appliances they might not hit the issue, or could siimply work around it and the admins never told the developers how they did it :D

**Comment by jamezp** [comment]
Not really related, but something I had just happened to be working on so I'll share here too. The Jakarta Servlet 6.1 spec requires RFC 6265 now, https://github.com/jakartaee/servlet/issues/37.

The [Jakarta REST spec](https://jakarta.ee/specifications/restful-ws/3.1/jakarta-restful-ws-spec-3.1) itself doesn't specify a specific RFC for the cookie, but it does indicate in the `jakarta.ws.rs.core.NewCookie` that it implements RFC-2109.

I don't think we can just remove the version, but I do think we could add a property which could control whether or not it gets added. I'll have to do some more research on whether we could fully support RFC 6265. I do think we should propose the REST spec should implement this though. I'll file an issue for this when I get some time.

**Comment by gsmet** [comment]
I have another idea: could we not include the version if it's the default i.e. `1`?

This wouldn't change the behavior AFAIK and would solve the issue at hand?

**Comment by vmuzikar** [comment]
> I have another idea: could we not include the version if it's the default i.e. `1`?

If I understand it correctly, according to the [RFC 2109](https://datatracker.ietf.org/doc/html/rfc2109#section-4.2.2), the version is required in any case – default or not. And it's for `Set-Cookie`, not `Set-Cookie2`. So if we're following RFC 2109, I'd say Quarkus is doing it correctly that it's including the version. But the question from my perspective is which of the RFCs it should follow.

**Comment by thomasdarimont** [comment]
I don't think that the Version=1; is that relevant to be honest. 
Neither google chrome:
![Image](https://github.com/user-attachments/assets/72e0aa8d-ece4-4fcd-b84e-ecbf07001799)
nor firefox:
![Image](https://github.com/user-attachments/assets/871a258f-6a56-4641-b2d8-25cfa668c01f)

I also couldn't find a site that currently uses the Version=1; attribute in the set-cookie header.

So I think it's veeeeery rarely used.

**Comment by vmuzikar** [comment]
@thomasdarimont I agree but if I understood it correctly, the problem is the [`NewCookie` Jakarta API](https://jakarta.ee/specifications/restful-ws/3.1/apidocs/jakarta.ws.rs/jakarta/ws/rs/core/newcookie) is hinting the old RFC 2109 should be followed.

**Comment by thomasdarimont** [comment]
Just checked and the new / upcoming `Jakarta RESTful Web Services 4.0` https://jakarta.ee/specifications/restful-ws/4.0/apidocs/jakarta.ws.rs/jakarta/ws/rs/core/newcookie also refers to the outdated / deprecated cookie RFC2109 from 1997 instead of the new RFC 6265.

**Comment by FroMage** [comment]
RFC 6265: April 2011

IMO, we should absolutely default to that RFC. Now, perhaps that will break some legacy code, but it would have to be legacy code that predates Quarkus' existence by almost a decade :-/

**Comment by gsmet** [comment]
My take on this is that we should drop the `Version` from Quarkus REST and see how it goes and then discuss with @jamezp to see what we do for RESTEasy and the spec. My guess is that they just never updated the spec.

**Comment by geoand** [comment]
That sounds very reasonable

**Comment by jamezp** [comment]
I need to make a proposal to update to the Jakarta REST spec about updating the spec. It's definitely out of date for sure. 

I think for RESTEasy we need to figure out a way to support both the obsolete RFC 2109 and the new RFC 6265. My initial thought will be to use a configuration option to be able to turn on support for RFC 6265. This would be similar to what Undertow does.

That said, the [`NewCookie`](https://jakarta.ee/specifications/restful-ws/3.1/apidocs/jakarta.ws.rs/jakarta/ws/rs/core/newcookie) is somewhat limited in what it can expose about an RFC 6265 cookie. For example there's not really a way to expose things like `Partitioned` as we can't set arbitrary attributes on the `NewCookie`.
